### PR TITLE
Gradient update in ADAM method to fix the issue #100

### LINKED ISSUE #100: Applying the ADAM method to the variables instead of the gradients
In the recent PR merged, the ADAM method was implemented but it has an issue that it applies the ADAM method to the gradients of the objective function instead of the variables themselves. The main change is that instead of this line where the gradients are being updated:
`c_grads -= alpha * torch.div(mhat_c, torch.sqrt(vhat_c) + epsilon)`
We have to have 
`c_grads = -alpha * torch.div(mhat_c, torch.sqrt(vhat_c) + epsilon)`
or something similar. This is because the update is happening later when we have 
`c += dt * c_grads`. 
This applies to all solvers (MF, DL, Langevin, and pumped-Langevin) that have ADAM method implemented.

### PR DESCRIPTION
**Pull request by MehmetC-1QBit**
### Description

>  To address [the issue ](https://github.com/1QB-Information-Technologies/ccvm/issues/100#issue-1897028068), two versions of the gradient update in [Adam](https://doi.org/10.48550/arXiv.1412.6980) methods of the solver classes were updated:
- version#1
grad_assigned
- version#2
grad_add_assigned

**To investigate further**
>  ADAM methods in all solvers (`MF, DL, Langevin`) were reflected the change.
>  Demo scripts were updated correspondingly.

**Minor fix:**
> - Fixed issue with `examples/ccvm_boxqp_plot.py`
> - Updated `examples/README.md`

is selected with `hyperparameters=dict( beta1=0.9,  beta2=0.999,  alpha=0.2,  which_adam="ASSIGN")`.
Here one can choose either `which_adam="ASSIGN"` or `which_adam="ADD_ASSIGN"` for the desired gradient update.
   
### Checklist

- [x] I have added documentation for new/changed functionality in this PR
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `main`

### CODE REVIEW

**Review by FarhadK-1QBit — APPROVED** [review]

**Review by Estella-Wang-1qbit — CHANGES_REQUESTED** [review]
Great job! 👍 I left some comment for your consideration. Most of them pertain to our coding guidelines and aesthetics. If you have any questions or need further clarification on my comments, please feel free to ask. Once these comments are addressed, I'll be ready to approve it.

File: ccvm_simulators/solvers/dl_solver.py [hunk 1]
```diff
@@ -582,18 +579,59 @@ def _solve_adam(
         beta1 = hyperparameters["beta1"]
         beta2 = hyperparameters["beta2"]
         epsilon = 1e-8
+         
+        # Compute bias corrected grads using 1st and 2nd moments 
+        # with element-wise division
+        def update_grads_with_moment2_assign(
+                gradc, grads, mhatc, vhatc, mhats, vhats
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/solvers/dl_solver.py:586]: Thanks for bringing this up. Before addressing your question I would like you to keep in mind that `_solve_adam()` methods in the sover classes are experimental and the redundancies including the one you pointed out will be addressed in later issues/PRs.

**Let's get back to your question:**  This function (`update_grads_with_moment2_assign` in line 585) and the other function (`update_grads_with_moment2_addassign` in line 593) are  conditionally called in between 665-668 based on the choice (lines 608-618). For the sake of a single call in line 665-668 we need to keep `c_grads` and  `s_grads` as an argument.  
I hope my answer addresses your question.

File: ccvm_simulators/solvers/dl_solver.py [hunk 2]
```diff
@@ -582,18 +579,59 @@ def _solve_adam(
...
+        def update_grads_with_moment2_addassign(
+                gradc, grads, mhatc, vhatc, mhats, vhats
+        ):
+            return (
+                gradc + alpha * torch.div(mhatc, torch.sqrt(vhatc) + epsilon),
+                grads + alpha * torch.div(mhats, torch.sqrt(vhats) + epsilon)
+            )
+        
+        # Compute bias corrected grads using only 1st moment
+        def update_grads_without_moment2_assign(gradc, grads, mhatc, mhats):
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/solvers/dl_solver.py:602]: I suppose my answer in the previous comment address this one as well

File: ccvm_simulators/solvers/dl_solver.py [hunk 3]
```diff
@@ -606,32 +644,36 @@ def _solve_adam(
...
                 vhat_c = v_c / beta2i
                 vhat_s = v_s / beta2i
-                
-                # Compute bias corrected grads using 1st and 2nd moments
-                # in the form of element-wise division
-                c_grads -= alpha * torch.div(mhat_c, torch.sqrt(vhat_c) + epsilon)
-                s_grads -= alpha * torch.div(mhat_s, torch.sqrt(vhat_s) + epsilon)
 
+                # Compute bias corrected grads
+                c_grads, s_grads =\
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/solvers/dl_solver.py:665]: You are right, I fixed it. 
BTW I started using black with this project. Bu I am having a hard time in understanding why black sometimes does not hit this time of line breaking.

File: ccvm_simulators/solvers/mf_solver.py [hunk 4]
```diff
@@ -608,13 +611,43 @@ def _solve_adam(
         beta2 = hyperparameters["beta2"]
         epsilon = 1e-8
         
+        # Compute bias corrected grads using 1st and 2nd moments 
+        # with element-wise division
+        def update_grads_with_moment2_assign(grads, mhat, vhat):
+            return alpha * torch.div(mhat, torch.sqrt(vhat) + epsilon)
+        
+        def update_grads_with_moment2_addassign(grads, mhat, vhat):
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/solvers/mf_solver.py:619]: My answer in for `dl_solver.py` also applies here.

File: examples/ccvm_boxqp_dl.py [hunk 5]
```diff
@@ -30,9 +30,14 @@
...
             instance=boxqp_instance,
-            solve_type = "Adam", # solve_type=None refers to default (original) solver
+            solve_type="Adam",  # solve_type=None refers to default (original) solver
             post_processor=None,
-            hyperparameters=dict(beta1=0.9, beta2=0.999, alpha=0.001),
+            hyperparameters=dict(
+                beta1=0.9,
+                beta2=0.999,
+                alpha=0.2,
+                which_adam="ASSIGN", # Temporary flag for selecting desired gradient update either "ASSIGN" or "ADD_ASSIGN" in the Adam method.
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply examples/ccvm_boxqp_dl.py]: Same as in previous comment

File: ccvm_simulators/solvers/dl_solver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Estella-Wang-1qbit) [reply ccvm_simulators/solvers/dl_solver.py]: This sounds good. Thanks for the detailed explanation.

File: ccvm_simulators/solvers/dl_solver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Estella-Wang-1qbit) [reply ccvm_simulators/solvers/dl_solver.py]: Hmm. I haven't had this issue. But I guess when you try running `black` again, it fixed the backslash issue right?

[the remark this answers is not in the document]
Comment (Estella-Wang-1qbit) [reply ccvm_simulators/solvers/dl_solver.py]: We can look into this deeper if this continue to bother you in the future.

**Review by Estella-Wang-1qbit — APPROVED** [review]
Thank you Mehmet. Great work! 👍

File: ccvm_simulators/solvers/langevin_solver.py [hunk 6]
```diff
@@ -513,26 +536,32 @@ def _solve_adam(
...
+                beta2i = 1.0 - beta2 ** (i + 1)
                 vhat_c = v_c / beta2i
-                
+
                 # Compute bias corrected grad using 1st and 2nd moments
                 # in the form of element-wise division
-                c_grads -= alpha * torch.div(mhat_c, torch.sqrt(vhat_c) + epsilon)
+                #===============================================================
+                # c_grads = alpha * torch.div(mhat_c, torch.sqrt(vhat_c) + epsilon)
+                #===============================================================
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/solvers/langevin_solver.py]: This was the statement that the method call just below this line replaced it. I should have removed it

**Pull request merged**